src/coop_vec.cpp

```diff
@@ -0,0 +1,724 @@
+    }
+
+    void *p = nullptr;
+    Ref tmp = steal(jitc_var_data(target, false, &p));
+    Ref target_ptr = steal(jitc_var_pointer(backend, p, tmp, 1));
+
+    Ref mask = steal(jitc_var_bool(backend, true));
+    mask = steal(jitc_var_mask_apply(mask, size));
+
+    Variable v;
+    v.kind = (uint32_t) VarKind::CoopVecAccum;
+    v.type = (uint32_t) VarType::Void;
+    v.size = std::max(size, jitc_var(mask)->size);
+    v.backend = (uint32_t) backend;
+    v.literal = offset;
+    v.symbolic = jitc_flag(JitFlag::SymbolicScope);
+    v.dep[0] = target_ptr;
+    v.dep[1] = index;
+    v.dep[2] = mask;
+    jitc_var_inc_ref(target_ptr);
+    jitc_var_inc_ref(index);
+    jitc_var_inc_ref(mask);
+
+    uint32_t result = jitc_var_new(v, true);
+    jitc_var_mark_side_effect(result);
+    return target.release();
+}
+
+uint32_t jitc_coop_vec_outer_product_accum(uint32_t target_,
+                                           uint32_t target_size,
+                                           const MatrixDescr *descr,
+                                           uint32_t a, uint32_t b) {
+    JitBackend backend;
+    VarType vt;
+    uint32_t size;
+
+    if (target_)
+        target_ = unwrap(target_);
+
+    {
+        const Variable *v_a = jitc_var(a),
+                       *v_b = jitc_var(b);
+
+        if (!v_a->coop_vec || !v_b->coop_vec || v_a->type != v_b->type)
+            jitc_raise("jit_coop_vec_outer_product_accum(): 'a' and 'b' must "
+                       "be cooperative vectors of a compatible type!");
+
+        backend = (JitBackend) v_a->backend;
+        vt = (VarType) v_a->type;
+        size = std::max(v_a->size, v_b->size);
+
+        if (backend == JitBackend::CUDA) {
+            if (vt != VarType::Float16)
+                jitc_raise(
+                    "jit_coop_vec_outer_product_accum(): this operation is "
+                    "restricted to float16 precision on the CUDA/OptiX backend.");
+            if (descr->layout != MatrixLayout::TrainingOptimal)
+                jitc_raise("jit_coop_vec_outer_product_accum(): the matrix "
+                           "must be in training-optimal layout!");
+        }
+    }
+
+    Ref target = borrow(target_);
+    if (!target) {
+        uint64_t z = 0;
+        target = steal(jitc_var_literal(backend, vt, &z, target_size, true));
+    } else {
+        const Variable *target_v = jitc_var(target);
+        // Copy-on-Write logic. See the same line in jitc_var_scatter() for details
+        if (target_v->ref_count != 2 && target_v->ref_count_stashed != 1)
+            target = steal(jitc_var_copy(target));
+
+        if ((VarType) target_v->type != vt)
+            jitc_raise("jit_coop_vec_outer_product_accum(): source/target "
+                       "buffers have an incompatible type!");
+    }
+
+    void *p = nullptr;
+    Ref tmp = steal(jitc_var_data(target, false, &p));
+    Ref target_ptr = steal(jitc_var_pointer(backend, p, tmp, 1));
```

Comment:
Does the current implementation support multiple `outer_product_accumulate` to the same matrix?
It's quite common to evaluate the same MLP multiple times in the same forward pass (e.g. taking finite difference gradients of a neural field), so in the backward pass there would be multiple outer products targeting the same matrix that we want to keep in the same kernel.

Reply:
Yes, that is perfectly legal. Is there something in this code fragment that made you think otherwise?

Reply:
I was just thinking that one `outer_product_accumulate()` op may count as a side effect to `target`. So then taking `jitc_var_data()` of pointer would trigger an evaluation of the pending side effect?
I didn't check for this carefully though, just wanted to make sure that it works.